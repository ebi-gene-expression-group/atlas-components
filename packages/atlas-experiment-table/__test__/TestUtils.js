import React from 'react'
import styled from 'styled-components'

// Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (min, max) => {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min)) + min //The maximum is exclusive and the minimum is inclusive
}

const TableCellDiv = styled.div`
  font-size: 13px;
  font-family: Helvetica, Arial, FreeSans, "Liberation Sans", sans-serif;
`

const tableHeader = [
  { type: `sort`, title: `Loaded date`, width: 140, dataParam: `lastUpdate` },
  { type: `search`, title: `species`, width: 200, dataParam: `species` },
  { type: ``, title: `experiment description`, width: 360, dataParam: `experimentDescription`, link: `experimentAccession`, resource: `experiments`, endpoint: `Results` },
  { type: `plain`, title: `experiment factors`, width: 260, dataParam: `experimentalFactors` },
  { type: `sort`, title: `Number of assays`, width: 160, dataParam: `numberOfAssays`, link: `experimentAccession`, resource: `experiments`, endpoint: `Experiment Design` },
]

const data = [
  {
    experimentType: `SINGLE`,
    experimentAccession: `E-EHCA-2`,
    experimentDescription: `Melanoma infiltration`,
    lastUpdate: `16-11-2018`,
    numberOfAssays: 6638,
    numberOfContrasts: 0,
    species: `Mus musculus`,
    kingdom: `animals`,
    experimentalFactors: [`single cell identifier`, `sampling site`,`time`],
  },
  {
    experimentType: `SINGLE`,
    experimentAccession: `E-GEOD-99058`,
    experimentDescription: `Single cell`,
    lastUpdate: `11-10-2018`,
    numberOfAssays: 250,
    numberOfContrasts: 0,
    species: `Mus musculus`,
    kingdom: `animals`,
    experimentalFactors: [`single cell identifier`]
  },
  {
    experimentType: `SINGLE`,
    experimentAccession: `E-MTAB-5061`,
    experimentDescription: `healthy individuals and type 2 diabetes patients`,
    lastUpdate: `11-10-2018`,
    numberOfAssays: 3514,
    numberOfContrasts: 0,
    species: `Homo sapiens`,
    kingdom: `plants`,
    experimentalFactors: [`single cell identifier`, `disease`],
  }
]

export {getRandomInt, TableCellDiv, tableHeader, data}
