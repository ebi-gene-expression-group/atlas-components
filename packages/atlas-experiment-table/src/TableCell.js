import React from 'react'
import PropTypes from 'prop-types'
import URI from 'urijs'

import { Table, Text } from 'evergreen-ui'

const TableCell = ({ dataRow, dataKey, image, linkTo, host, width }) => {
  let cellItem = null
  if (image) {
    // If the column is an image put the image in the cell (or an unknown icon if the type is undefined)
    if (image[dataRow[dataKey]]) {
      cellItem = <img {...image[dataRow[dataKey]]}/>
    }
    else {
      cellItem = `❔`
    }
  } else if (Array.isArray(dataRow[dataKey])) {
    // If the contents of the cell is an array expand it to a list
    cellItem =
      <ul>
        {dataRow[dataKey].map(
          (element, index) =>
            <li key={`${index}`}>
              {element}
            </li>
        )}
      </ul>
  } else {
    // Any other type (i.e. string, number) put it as-is in the cell
    cellItem = dataRow[dataKey]
  }
  
  // Evergreen’s Table.TextCell ellipsifies content by default and we don’t want that
  return (
    <Table.Cell flexGrow={width}>
      <Text size={400}>
        {
          linkTo ?
            <a href={URI(linkTo(dataRow), host).toString()}>{cellItem}</a> :
            cellItem
        }
      </Text>
    </Table.Cell>
  )
}

TableCell.propTypes = {
  dataRow: PropTypes.object.isRequired,
  dataKey: PropTypes.string.isRequired,
  image: PropTypes.objectOf(
    PropTypes.shape({
      src: PropTypes.string,
      alt: PropTypes.string,
      title: PropTypes.string
    })
  ),
  host: PropTypes.string,
  linkTo: PropTypes.func,
  width: PropTypes.number,
}

TableCell.defaultProps = {
  image: null,
  linkTo: null,
  host: ``,
  width: 1,
}

export default TableCell
