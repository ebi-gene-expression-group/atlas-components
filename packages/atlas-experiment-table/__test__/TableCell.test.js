import React from 'react'
import { shallow, mount } from 'enzyme'

import { Table, Text } from 'evergreen-ui'

import TableCell from '../src/TableCell'

import randomString from 'random-string'
import { getRandomInt } from './TestUtils'

describe(`TableCell`, () => {
  const MAX_DATA_ROW_ARRAY_LENGTH = 20

  const props = {
    dataRow: {
      string: randomString(),
      experimentAccession: randomString(),
      number: getRandomInt(0, 9001),
      arrayOfStrings:
        Array.apply(0, Array(getRandomInt(1, MAX_DATA_ROW_ARRAY_LENGTH)))
          .map(() => randomString()),
      arrayOfNumbers:
        Array.apply(0, Array(getRandomInt(1, MAX_DATA_ROW_ARRAY_LENGTH)))
          .map(() => getRandomInt(0, 9001)),
      arrayOfStringsAndNumbers:
        Array.apply(0, Array(getRandomInt(1, MAX_DATA_ROW_ARRAY_LENGTH)))
          .map(() => Math.random() < 0.5 ? randomString() : getRandomInt(0, 9001))
    }
  }

  test(`is a plain text field if keyed row value is a string`, () =>{
    const wrapper = shallow(<TableCell {...props} dataKey={`string`}/>)

    expect(wrapper.find(Table.Cell).find(Text).children()).toHaveText(props.dataRow.string)
  })

  test(`is a plain text field if keyed row value is a number`, () =>{
    const wrapper = shallow(<TableCell {...props} dataKey={`number`}/>)

    expect(wrapper.find(Table.Cell).find(Text).children()).toHaveText(props.dataRow.number.toString())
  })

  test.each(
    [`arrayOfStrings`, `arrayOfNumbers`, `arrayOfStringsAndNumbers`]
  )(`is an unordered-list if keyed row value is an array`, (arrayDataKey) => {
    const wrapper = shallow(<TableCell {...props} dataKey={arrayDataKey}/>)

    expect(wrapper.find(Table.Cell).find(`li`)).toHaveLength(props.dataRow[arrayDataKey].length)
    expect(wrapper.find(Table.Cell).find(`li`).first()).toHaveText(props.dataRow[arrayDataKey][0].toString())
    expect(wrapper.find(Table.Cell).find(`li`).last())
      .toHaveText(props.dataRow[arrayDataKey][props.dataRow[arrayDataKey].length - 1].toString())
  })

  test(`is an image if a dictionary of images is passed as prop`, () => {
    const image = {
      [props.dataRow.string]: {
        src: `imageUrl`,
        alt: `altText`
      }
    }
    const wrapper = shallow(<TableCell {...props} image={image} dataKey={`string`}/>)

    expect(wrapper.find(Table.Cell)).toContainExactlyOneMatchingElement(`img`)
    expect(wrapper.find(Table.Cell).find(`img`)).toHaveProp({src: `imageUrl`, alt: `altText`})
  })

  test.each(
      [`E-ANND-12`, `testExperimentAccession`, `E-MTAB-12`]
  )(`contains an icon and check if it is an Anndata experiment or no icons for others`, (arrayDataKey) => {
    props.dataRow.experimentAccession = arrayDataKey

    const wrapper = shallow(<TableCell {...props} label={`Title`} dataKey={`string`} />)
    0
    if (arrayDataKey.startsWith(`E-ANND`)) {
      // If it's an Anndata experiment, there should be an icon
      expect(wrapper.find('.icon').length).toBe(1);
    } else {
      // For other types, there should be no icon
      expect(wrapper.find('.icon').length).toBe(0);
    }
  })

  test(`defaults to a fallback icon if a suitable image can’t be found`, () => {
    const image = {}
    const wrapper = shallow(<TableCell {...props} image={image} dataKey={`string`}/>)

    expect(wrapper.find(Table.Cell).find(`img`)).not.toExist()
    expect(wrapper.find(Table.Cell).find(Text).children()).toHaveText(`❔`)
  })

  test(`shows a link if a linkTo function has been passed`, () => {
    const linkTo = dataRow => `https://some.domain/${dataRow.string}/someResource`
    const wrapper = shallow(<TableCell {...props} linkTo={linkTo} dataKey={`number`}/>)

    expect(wrapper.find(Table.Cell)).toContainExactlyOneMatchingElement(`a`)
    expect(wrapper.find(Table.Cell).find(`a`)).toHaveProp({ href: linkTo(props.dataRow) })
  })

  test(`matches snapshot (sring)`, () => {
    const dataRow = {
      string: `foobar`
    }
    const wrapper = mount(<TableCell dataRow={dataRow} dataKey={`string`}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (number)`, () => {
    const dataRow = {
      number: 9001
    }
    const wrapper = mount(<TableCell dataRow={dataRow} dataKey={`number`}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (array of strings)`, () => {
    const dataRow = {
      arrayOfStrings: [
        `I mean, why would a Pop-Tart want to live inside a toaster, Rick?`,
        `I mean, that would be like the scariest place for them to live.`,
        `You know what I mean?`]
    }
    const wrapper = mount(<TableCell dataRow={dataRow} dataKey={`arrayOfStrings`}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (array of numbers)`, () => {
    const dataRow = {
      arrayOfNumbers: [42, 9001]
    }
    const wrapper = mount(<TableCell dataRow={dataRow} dataKey={`arrayOfNumbers`}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (array of strings and numbers)`, () => {
    const dataRow = {
      arrayOfStringsAndNumbers: [
        `Answer to the ultimate question of life, the Universe, and everything`,
        42,
        `Vegeta, what does the scouter say about his power level?`,
        9001]
    }
    const wrapper = mount(<TableCell dataRow={dataRow} dataKey={`arrayOfStringsAndNumbers`}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (image)`, () => {
    const dataRow = {
      string: `foobar`
    }
    const image={
      [dataRow.string]: {
        src: `https://imgs.xkcd.com/comics/ringtone_timeline_2x.png`,
        alt: `Ringtone Timeline`,
        title: `No one likes my novelty ringtone, an audio recording of a phone on vibrate sitting on a hard surface.`
      }
    }
    const wrapper = mount(<TableCell dataRow={dataRow} image={image} dataKey={`string`}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (missing image)`, () => {
    const dataRow = {
      string: `foobar`
    }
    const image = {}
    const wrapper = mount(<TableCell dataRow={dataRow} image={image} dataKey={`string`}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (text link)`, () => {
    const dataRow = {
      string: `foobar`
    }
    const toLink = dataRow => `https://go.to/${dataRow.string}`
    const wrapper = mount(<TableCell dataRow={dataRow} toLink={toLink} dataKey={`string`}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (image link)`, () => {
    const dataRow = {
      string: `foobar`
    }
    const image={
      [dataRow.string]: {
        src: `https://imgs.xkcd.com/comics/ringtone_timeline_2x.png`,
        alt: `Ringtone Timeline`,
        title: `No one likes my novelty ringtone, an audio recording of a phone on vibrate sitting on a hard surface.`
      }
    }
    const toLink = dataRow => `https://go.to/${dataRow.string}`
    const wrapper = mount(<TableCell dataRow={dataRow} image={image} toLink={toLink} dataKey={`string`}/>)
    expect(wrapper).toMatchSnapshot()
  })
})
