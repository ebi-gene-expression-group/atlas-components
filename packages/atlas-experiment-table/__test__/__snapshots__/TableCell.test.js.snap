// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableCell matches snapshot (array of numbers) 1`] = `
<TableCell
  dataKey="arrayOfNumbers"
  dataRow={
    {
      "arrayOfNumbers": [
        42,
        9001,
      ],
    }
  }
  host=""
  image={null}
  linkTo={null}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                <ul>
                  <li
                    key="0"
                  >
                    42
                  </li>
                  <li
                    key="1"
                  >
                    9001
                  </li>
                </ul>
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;

exports[`TableCell matches snapshot (array of strings and numbers) 1`] = `
<TableCell
  dataKey="arrayOfStringsAndNumbers"
  dataRow={
    {
      "arrayOfStringsAndNumbers": [
        "Answer to the ultimate question of life, the Universe, and everything",
        42,
        "Vegeta, what does the scouter say about his power level?",
        9001,
      ],
    }
  }
  host=""
  image={null}
  linkTo={null}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                <ul>
                  <li
                    key="0"
                  >
                    Answer to the ultimate question of life, the Universe, and everything
                  </li>
                  <li
                    key="1"
                  >
                    42
                  </li>
                  <li
                    key="2"
                  >
                    Vegeta, what does the scouter say about his power level?
                  </li>
                  <li
                    key="3"
                  >
                    9001
                  </li>
                </ul>
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;

exports[`TableCell matches snapshot (array of strings) 1`] = `
<TableCell
  dataKey="arrayOfStrings"
  dataRow={
    {
      "arrayOfStrings": [
        "I mean, why would a Pop-Tart want to live inside a toaster, Rick?",
        "I mean, that would be like the scariest place for them to live.",
        "You know what I mean?",
      ],
    }
  }
  host=""
  image={null}
  linkTo={null}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                <ul>
                  <li
                    key="0"
                  >
                    I mean, why would a Pop-Tart want to live inside a toaster, Rick?
                  </li>
                  <li
                    key="1"
                  >
                    I mean, that would be like the scariest place for them to live.
                  </li>
                  <li
                    key="2"
                  >
                    You know what I mean?
                  </li>
                </ul>
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;

exports[`TableCell matches snapshot (image link) 1`] = `
<TableCell
  dataKey="string"
  dataRow={
    {
      "string": "foobar",
    }
  }
  host=""
  image={
    {
      "foobar": {
        "alt": "Ringtone Timeline",
        "src": "https://imgs.xkcd.com/comics/ringtone_timeline_2x.png",
        "title": "No one likes my novelty ringtone, an audio recording of a phone on vibrate sitting on a hard surface.",
      },
    }
  }
  linkTo={null}
  toLink={[Function]}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                <img
                  alt="Ringtone Timeline"
                  src="https://imgs.xkcd.com/comics/ringtone_timeline_2x.png"
                  title="No one likes my novelty ringtone, an audio recording of a phone on vibrate sitting on a hard surface."
                />
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;

exports[`TableCell matches snapshot (image) 1`] = `
<TableCell
  dataKey="string"
  dataRow={
    {
      "string": "foobar",
    }
  }
  host=""
  image={
    {
      "foobar": {
        "alt": "Ringtone Timeline",
        "src": "https://imgs.xkcd.com/comics/ringtone_timeline_2x.png",
        "title": "No one likes my novelty ringtone, an audio recording of a phone on vibrate sitting on a hard surface.",
      },
    }
  }
  linkTo={null}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                <img
                  alt="Ringtone Timeline"
                  src="https://imgs.xkcd.com/comics/ringtone_timeline_2x.png"
                  title="No one likes my novelty ringtone, an audio recording of a phone on vibrate sitting on a hard surface."
                />
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;

exports[`TableCell matches snapshot (missing image) 1`] = `
<TableCell
  dataKey="string"
  dataRow={
    {
      "string": "foobar",
    }
  }
  host=""
  image={{}}
  linkTo={null}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                ❔
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;

exports[`TableCell matches snapshot (number) 1`] = `
<TableCell
  dataKey="number"
  dataRow={
    {
      "number": 9001,
    }
  }
  host=""
  image={null}
  linkTo={null}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                9001
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;

exports[`TableCell matches snapshot (sring) 1`] = `
<TableCell
  dataKey="string"
  dataRow={
    {
      "string": "foobar",
    }
  }
  host=""
  image={null}
  linkTo={null}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                foobar
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;

exports[`TableCell matches snapshot (text link) 1`] = `
<TableCell
  dataKey="string"
  dataRow={
    {
      "string": "foobar",
    }
  }
  host=""
  image={null}
  linkTo={null}
  toLink={[Function]}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Text)
            size={400}
          >
            <Box
              boxSizing="border-box"
              className=""
              color="#425A70"
              fontFamily=""SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol""
              fontSize="14px"
              fontWeight={400}
              is="span"
              letterSpacing="-0.05px"
              lineHeight="20px"
            >
              <span
                className="ub-color_425A70 ub-fnt-fam_b77syt ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-box-szg_border-box"
              >
                foobar
              </span>
            </Box>
          </ForwardRef(Text)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</TableCell>
`;
